Failover framework

thed636 opened the pull request:
Failover micro framework
===========

Introduce failover mechanism with retry strategy as a first implemented strategy.

RATIONALE
----------

Distributed environment is unstable, so we want to have recovery mechanisms. Failover micro-framework is created to provide such customisable and easy to use solution. Here we have two new concepts: *Strategy* and *Try*.

* **Strategy** is entity which determines how to try to recover an operation which failed. 
* **Try** is an entity which holds a context of an operation and decides should we try to recover a failed operation or give an error for a user.

EXTENSION
----------

The `ozo::fallback::retry` strategy is an example of the framework extension.

OPERATION SUPPORT FAILOVER
-----------------

* request
* execute
* get_connection
* begin

BASIC USAGE EXAMPLE
----------------

Operation with failover looks like this:

```c++
execute[fallback::retry(connection_error)*3](conn_info[io], query, 3s, yield);
```

see [retry_request.cpp](https://github.com/yandex/ozo/pull/129/files#diff-0faf5b07ab5932bd18487a1b85009ef6) for more details.

yagzalex commented on include/ozo/detail/cancel_timer_handler.h:
```diff
@@ -30,4 +31,25 @@ struct cancel_timer_handler {
+    }
+};
+
+template <typename TimeConstrain, typename Handler>
```
Did you mean `TimeConstraint`?

thed636 replied:
Yep, thanks.

Commit:
Update Handler documentation

Commit:
Fallback micro framework

Introduce fallback mechanism with retry strategy as a first implemented
strategy.

RATIONALE

Distributed environment is unstable, so we want to have recovery mechanisms.
Fallback microframework is created to provide such customizeable and easy
to use solution. Here we have two new concepts: Strategy and Try.

Strategy is entity which determines how to try to recover an operation which
failed. Try is an entity which holds a context of an operation and decides
should we try to recover a failed operation or give an error for a user.

OPERATION SUPPORT FALLBACK

 * request
 * execute
 * get_connection
 * begin

BASIC USAGE

auto conn = execute[fallback::retry(connection_error)*3](conn_info[io], query, 3s, yield[ec]);

yagzalex commented on include/ozo/failover/retry.h:
```diff
@@ -0,0 +1,281 @@
+    ozo::request[retry*3](pool, query, .5s, out, yield);
+     * @endcode
+     */
+    auto times(int n) const {
```
Maybe you should named that as 'tries'. The name 'times' is really confusing, cause it can be misunderstood as time constraints or something.

thed636 replied:
**Done**

yagzalex commented on tests/failover/retry.cpp:
```diff
@@ -0,0 +1,191 @@
+}
+
+TEST_F(get_try_time_constraint, should_return_zero_time_left_for_try_count_zero) {
+    EXPECT_EQ(duration{0}, ozo::failover::detail::get_try_time_constraint(3s, 0, now));
```
Don't you mean to use 'deadline' from here? Otherwise, I don't see a big difference with 'should_return_zero_duration_for_try_count_zero' test.

thed636 replied:
**Done**
Yeah, thanks!

yagzalex commented on include/ozo/failover/retry.h:
```diff
@@ -0,0 +1,281 @@
+     * @param n --- number of tries
+     * @return `retry_strategy` specialization object
+     */
+    auto operator * (int n) const { return times(n); }
```
Crazy thought. I think it would be handy to also have
```
template <typename ErrorConditions>
auto operator * (int n, retry_strategy<ErrorConditions> rs);
```
Because when you write `ozo::request[ozo::failover::retry(ozo::errc::network) * 3](...)` it is a little bit hard to find the exact counts of tries.
And this will make it easier `ozo::request[3 * ozo::failover::retry(ozo::errc::network)](...)`

thed636 replied:
**Done**
